### PR TITLE
Spark: Implement an action to remove orphan files

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds a Spark action that removes orphan data and metadata files that can be left in some edge cases like executor preemption.

### CODE REVIEW

File: spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+    Dataset<Row> resultDF = spark.read().format("iceberg").load(tableLocation);
+    List<ThreeColumnRecord> actualRecords = resultDF
+        .as(Encoders.bean(ThreeColumnRecord.class))
+        .collectAsList();
+    Assert.assertEquals("Rows must match", expectedRecords, actualRecords);
+  }
+
+  @Test
+  public void testAllValidFilesAreKept() throws IOException, InterruptedException {
```
Comment (prodeezy) [inline spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: Can you also add a test for Write-Audit-Publish (WAP) workflow case where a snapshot can be staged (using the cherrypicking operation), where it's not part of the list of active snapshots. So expected behavior should be that this action should not delete those staged files as orphan files.
There are tests in `TestWapWorkflow` that illustrate this case.

Comment (aokolnychyi) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: Good idea, will add a case for that.

File: api/src/main/java/org/apache/iceberg/io/LocationProvider.java [hunk 2]
```diff
@@ -29,6 +29,12 @@
  * Implementations must be {@link Serializable} because instances will be serialized to tasks.
  */
 public interface LocationProvider extends Serializable {
+
+  /**
+   * Return a fully-qualified data location.
+   */
+  String dataLocation();
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/LocationProvider.java:36]: What about providers that don't have a reliable location? Do they return null?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    return new RemoveOrphanFilesActionResult(orphanDataFiles);
+  }
+
+  private Dataset<Row> buildValidDataFileDF() {
+    return spark.read().format("iceberg")
+        .load(allDataFilesTable)
+        .select("file_path")
+        .distinct();
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:116]: You probably don't need `distinct` if you're using this in a left anti-join. That just introduces an additional shuffle.

File: spark/src/main/java/org/apache/iceberg/Actions.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  public static Actions forTable(SparkSession spark, Table table) {
+    return new Actions(spark, table);
+  }
+
+  public static Actions forTable(Table table) {
+    return new Actions(SparkSession.active(), table);
+  }
+
+  public RemoveOrphanFilesAction removeOrphanFiles() {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/Actions.java:42]: Are you planning to make `removeOrphanMetadataFiles` as well, or would that be added to this?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.List;
+
+public class RemoveOrphanFilesActionResult {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java:24]: Why return a result class when you could just return `List<String>`? Do we need these wrappers?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    this.allDataFilesTable = newAllDataFilesTable;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction olderThan(long newOlderThanTimestamp) {
+    this.olderThanTimestamp = newOlderThanTimestamp;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction dryRun(boolean newDryRun) {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:83]: In other places, we use `deleteWith` instead of a dry run flag. Should we do that for consistency here?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  RemoveOrphanFilesAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.hadoopConf = new SerializableConfiguration(spark.sessionState().newHadoopConf());
+    this.fileIO = table.io();
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.dataLocation = table.locationProvider().dataLocation();
+  }
+
+  public RemoveOrphanFilesAction allDataFilesTable(String newAllDataFilesTable) {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:73]: This seems awkward, but I'm not sure a better way to do it.

File: api/src/main/java/org/apache/iceberg/io/LocationProvider.java (hunk 2 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/LocationProvider.java:36]: Instead of requiring a data location, why not use `table.location()`?

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/Actions.java:42]: We should add Javadoc here that explains what happens. That should note that the table location needs to be accessible for listing via Hadoop FileSystem, but the table's FileIO will be used to delete.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path dataPath = new Path(dataLocation);
+      FileSystem fs = dataPath.getFileSystem(hadoopConf.value());
+
+      for (FileStatus file : fs.listStatus(dataPath, HiddenPathFilter.get())) {
+        // TODO: handle custom metadata folders
+        // we need to ignore the metadata folder when data is written to the root table location
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:129]: Do we?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        if (file.isDirectory() && !"metadata".equals(file.getPath().getName())) {
+          topLevelDirs.add(file.getPath().toString());
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:131]: I would probably opt not to do this check. Instead, there is little cost to doing the anti-join with all known data and metadata file paths. I'd probably do that instead of worrying about metadata location.

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: Both options would work. In order to keep the actions API smaller, I think we can extend the existing API with more properties.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java (hunk 5 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java:24]: That's something we need to discuss. The idea behind this was to report deleted metadata and data files separately when we extend this API in the future. If we want to report all deleted files together or if we want to provide a separate action for metadata, we can return `List<String>`. 

Since it is a new API, we need to decide what actions should report back. For example, do actions need to return a summary? In `RewriteManifestsAction`, this can be a list of manifests that were replaced and a list of manifests that were added. People can infer that themselves but it will require more steps. They would need to assign a label to the produced snapshot, load that snapshot and then access the snapshot summary which will contain only numbers.

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: The name is generic enough to cover both data and metadata files.

File: api/src/main/java/org/apache/iceberg/io/LocationProvider.java (hunk 2 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/io/LocationProvider.java:36]: My assumption was that there would always be at least some base reliable data location. Is that assumption false?

Technically, data location can be arbitrary today and might be outside `table.location()`. That would be odd but possible.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 6 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:83]: Makes sense.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java (hunk 5 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java:24]: If this will return two lists in the future, then I think it makes sense to keep the class. Thanks!

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 7 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:73]: Any other ideas are more than welcome :)

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 8 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:129]: It is a leftanti join between all files and valid data files. If we don't ignore the metadata folder, we will remove all metadata files, right?

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:129]: I think we want better protection for the metadata files. We could have a case where the user put metadata files into the data folder and then relocated the config to a different place. We still want to protect the metadata.

I think that means that we should do the anti-join with all data and metadata file paths. We can get recent metadata files from:
* The `all_manifests` table has manifest locations
* The `snapshots` table has manifest list locations
* The table metadata has the current metadata location and a list of old metadata locations

I think it should be sufficient to protect all of those paths. Then we don't need to worry about a prefix to protect metadata.

File: api/src/main/java/org/apache/iceberg/io/LocationProvider.java (hunk 2 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/LocationProvider.java:36]: I think we should clean `table.location()` by default and allow the user to configure a different path. That covers all reasonable cases.

It should be safe to clean `table.location()` because we need to protect all metadata file paths individually anyway, instead of using a prefix. So we may as well clean the whole table location.

Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/io/LocationProvider.java:36]: Yeah, that's what I was thinking as well. If we spend a bit more effort to also collect all metadata files, we can clean the table location completely.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 7 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:73]: In other places, we detect path tables by checking `contains("/")`. We could do that here to construct the metadata table names:

```
public String metadataTableName(Table table, String metaTable) {
  String tableName = table.toString()
  if (tableName.contains("/")) {
    return tableName + "#" + metaTable;
  } else {
    return tableName + "." + metaTable;
  }
}
```

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 8 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:129]: Since we limit the number of previous metadata locations we keep in each new metadata file, the list of old metadata locations might not be complete in the current metadata file. It seems safe to remove others (assuming they are old enough). What do you think, @rdblue?

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:129]: I think it's a good thing to remove them.

File: spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java (hunk 1 above)
Comment (rdblue) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: I agree it would be nice to have a test for it. This should work because the metadata tables used return all files reachable by metadata, not just the ones in a single snapshot. We use the same table for a similar check in our environment.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 7 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:73]: This almost works. Unfortunately, `table.toString()` might return anything. For example, it will prepend `hive.` for friendly names in the Hive catalog and we won't be able to resolve `hive.db.table.all_data_files`. Exposing a correct `TableIdentifier` in `Table` would help but that would mean modifying public `BaseTable` as well.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:73]: Okay, I thought this would work because it is what we do in our Spark 2.4 build. Since we are using DSv2 catalogs, when the catalog adds its name to the identifier we actually get a working multi-catalog identifier.

Maybe we should add something to remove `hive.` for now, and take it out for Spark 3.0.

File: spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: There is one case when `all_data_file_table` won't report WAP files: when there is no snapshot.

```
  @Override
  public CloseableIterable<FileScanTask> planFiles() {
    Snapshot snapshot = snapshot();
    if (snapshot != null) {
      LOG.info("Scanning table {} snapshot {} created at {} with filter {}", table,
          snapshot.snapshotId(), formatTimestampMillis(snapshot.timestampMillis()),
          rowFilter);

      Listeners.notifyAll(
          new ScanEvent(table.toString(), snapshot.snapshotId(), rowFilter, schema()));

      return planFiles(ops, snapshot, rowFilter, caseSensitive, colStats);

    } else {
      LOG.info("Scanning empty table {}", table);
      return CloseableIterable.empty();
    }
  } 
```

Comment (aokolnychyi) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: Also, `table.currentSnapshot().manifestListLocation()` in `location()`can lead to a NPE.

Comment (aokolnychyi) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:166]: I've created #904 and #905 to address later.

Comment (aokolnychyi) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:164]: As long as we have a current snapshot, it does work correctly. I've added a test.

File: api/src/main/java/org/apache/iceberg/io/LocationProvider.java (hunk 2 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/io/LocationProvider.java]: Done.

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: It cleans both now.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java (hunk 5 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesActionResult.java]: I removed the wrapper as it would require additional complexity to report data and metadata files separately. Users can call the action on data and metadata separately if they need such functionality.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 8 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Added a test for old metadata files.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 9 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:164]: Done.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Done.

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: Will add docs in the next pass. If we decide to extend `FileIO` with list operations too, we should migrate to that here.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceHiveTables.java [hunk 10]
```diff
@@ -78,12 +81,12 @@ public static void stopMetastoreAndSpark() {
   }
 
   @After
-  public void dropTable() throws Exception {
-    clients.run(client -> {
-      client.dropTable(TestIcebergSourceHiveTables.currentIdentifier.namespace().level(0),
-          TestIcebergSourceHiveTables.currentIdentifier.name());
-      return null;
-    });
+  public void dropTable() throws IOException {
```
Comment (aokolnychyi) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceHiveTables.java:84]: We have to clean the location properly.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  private String metadataTableName(MetadataTableType type) {
+    String tableName = table.toString();
+    if (tableName.contains("/")) {
+      return tableName + "#" + type;
+    } else {
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:235]: I could probably add a condition to this branch: `tableName.startsWith("hadoop.") || tableName.startsWith("hive.")`. Also, we will need a comment here.

**Comment by mehtaashish23** [comment]
Thanks @aokolnychyi for the PR. Based on what I see on PR, you are trying to clean up files that are not referenced by `metadata` table and `all_data_files` table. I am actually going to verify it, as I get to use this work in our project, but do the files which fail to get deleted during cleanup after commit [here], is also cleaned up with this work? My point is whether the dataFile which is deleted as part of an expired snapshot belongs to `all_data_files` table or not.

[here]: https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/RemoveSnapshots.java#L145

**Comment by aokolnychyi** [comment]
@mehtaashish23, this action should remove all orphan files including those that we failed to delete while expiring snapshots. The `all_data_files` metadata table gives us currently referenced files in metadata.

File: spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java (hunk 1 above)
Comment (rdblue) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:164]: I thought we addressed the metadata table problem when there is no current snapshot in #801. I'll check to see why that doesn't work. My initial guess is that PR refers to static tables and this is a parallel table.

Comment (rdblue) [reply spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java:164]: It's just individual tables that were fixed. The solution is to override `planFiles`:

```java
    @Override
    public CloseableIterable<FileScanTask> planFiles() {
      // override planFiles to avoid the check for a current snapshot because this metadata table is for all snapshots
      return CloseableIterable.withNoopClose(HistoryTable.this.task(this));
    }
```

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: Hm. I've been avoiding a list operation in `FileIO`. Maybe we could add it as an optional trait? I just don't want that interface to get huge and complicated.

File: spark/src/main/java/org/apache/iceberg/Action.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+public interface Action<R> {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/Action.java:22]: Should we create an `actions` package?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 11 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:235]: +1 for a comment

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    List<String> otherMetadataFiles = Lists.newArrayList();
+
+    for (Snapshot snapshot : table.snapshots()) {
+      String manifestListLocation = snapshot.manifestListLocation();
+      if (manifestListLocation != null) {
+        otherMetadataFiles.add(manifestListLocation);
+      }
+    }
+
+    otherMetadataFiles.add(ops.metadataFileLocation("version-hint.text"));
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:139]: Good catch!

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+
+    return manifestDF.union(otherMetadataFileDF);
+  }
+
+  private Dataset<Row> buildActualFileDF() {
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path path = new Path(location);
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:159]: Now that this is the table's location, we expect this to contain just two directories: `data` and `metadata`. I think the intent of this methods was to parallelize on the first level of partition directories, but that's not what will happen here.

It's a bit more tricky because we don't know the convention actually matches the default structure, but I think it would be reasonable to traverse the first 2 layers of directories to build the top-level set. To do that, adding a `depth` parameter to the recursive traversal makes sense so you can use it here and return after 2 levels (or a configurable number). That would also be a good thing for the parallel traversal to ensure this won't get caught in a symlink loop.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    return spark.createDataset(matchingFileRDD.rdd(), Encoders.STRING()).toDF("file_path");
+  }
+
+  private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
+      Broadcast<SerializableConfiguration> conf,
+      long olderThanTimestamp) {
+
+    return (FlatMapFunction<Iterator<String>, String>) dirs -> {
+      List<String> files = Lists.newArrayList();
+      Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:195]: Nit: `predicate` isn't a very descriptive name. Maybe `pastOperationTimeLimit` instead?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+      for (FileStatus file : fs.listStatus(path, HiddenPathFilter.get())) {
+        if (file.isDirectory()) {
+          childDirs.add(file.getPath().toString());
+        } else if (file.isFile() && predicate.test(file)) {
+          matchingFiles.add(file.getPath().toString());
+        }
+      }
+
+      for (String childDir : childDirs) {
+        List<String> childDirFiles = listDirRecursively(childDir, predicate, conf);
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:221]: What about passing a result list into this method to avoid creating lots of small lists and merging them together in every call?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
+    this.deleteFunc = newDeleteFunc;
+    return this;
+  }
+
+  @Override
+  public List<String> execute() {
+    Preconditions.checkArgument(olderThanTimestamp != null, "olderThanTimestamp must be set");
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:96]: Can we default it to `System.currentTimeMillis() - THREE_DAYS_MS`?

**Comment by rdblue** [comment]
@aokolnychyi, this looks almost ready to me. The parallel file listing looks like it needs to be updated, and we need javadoc. Otherwise I think the other points can be done as follow-ups.

File: spark/src/main/java/org/apache/iceberg/Actions.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: I would not worry about that for now. We will need to expose some sort of `FileStatus` too if we decide to support list operations. Let's first have that use case when the location cannot be listed by Hadoop `FileSystem` and requires a custom `FileIO` implementation.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/Actions.java:42]: I agree

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 14 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: There is one more problem: the initial number of locations might be pretty small. It seems beneficial to list, for example, 3 levels on the driver and then parallelize. The number of top-level partitions might be small. At the same time, we should avoid listing too much on the driver if the data is written to the root table location. That's why I modified the listing logic so that we list 3 levels by default by don't list locations that have more than 10 sub-locations. The latter ones will be listed in a distributed manner. This should cover cases with a lot of top-level and leaf partitions.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 11 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Added.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    Dataset<Row> otherMetadataFileDF = spark
+        .createDataset(otherMetadataFiles, Encoders.STRING())
+        .toDF("file_path");
+
+    return manifestDF.union(otherMetadataFileDF);
+  }
+
+  private Dataset<Row> buildActualFileDF() {
```
Comment (prodeezy) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:184]: a comment here on criteria for building the list of actual files would be helpful.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 16 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Done.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 17 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Done.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 15 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: I kept it to stay on one line below.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 14 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: Actually, ignore my previous comment. I'll think more about this.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: I think it's fine to list a fixed-number of levels. If all the data is written to the root location, there's nothing we can do anyway because it can't be parallelized.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final SerializableConfiguration hadoopConf;
+  private final int partitionDiscoveryParallelism;
+  private final Table table;
+  private final TableOperations ops;
+
+  private String location = null;
+  private long olderThanTimestamp = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(3);
+  private Consumer<String> deleteFunc = new Consumer<String>() {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:76]: Nit: we should be able to write this a `table.io()::deleteFile`

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 14 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: After thinking about this more, I think it still makes sense to list 2 levels and stop whenever we hit let's say 10 sub-locations. It won't solve the problem when the number of top-level partitions is small. However, it should help if the table is partitioned but the data is written to the table location. Consider tables that were migrated from Hive. If they have 1000 top-level partitions and 10 sub-partitions, we will be listing 10000 locations on the driver only for the first two levels.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 18 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:184]: @prodeezy, do you mean the actual algorithm or that we select files older than a given timestamp?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 14 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java]: @rdblue, what do you think?

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+      Broadcast<SerializableConfiguration> conf,
+      long olderThanTimestamp) {
+
+    return (FlatMapFunction<Iterator<String>, String>) dirs -> {
+      List<String> subDirs = Lists.newArrayList();
+      List<String> files = Lists.newArrayList();
+
+      Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
+
+      int maxDepth = Integer.MAX_VALUE;
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:269]: This seems excessive, but not really that dangerous. When listing in executors, the purpose is to exit even if there is a reference cycle in the file system. This would technically do that, but would recurse 2 billion levels so the more likely failure is a stack overflow.

That's alright since it's the behavior that was here before, but I think it would be better to set this to 2,000 or something large but reasonable and then throw an exception if there are remaining directories when it returns.

File: spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java (hunk 19 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:76]: That's what I tried in the first place. Unfortunately, it complains with `Variable 'table' might not have been initialized`. I think we had the same problem in `RemoveSnapshots`.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java:76]: That makes sense. Let's go with this then.

**Comment by rdblue** [comment]
I'm merging this since it's large and the remaining comments are minor. That avoids needing to re-read the whole commit for small updates. Thanks for adding this, @aokolnychyi! I think it is going to be really useful.
